File: packages/sdk-python/fourmica_sdk/client/withdraw.py
"""Withdrawing collateral. A request opens a waiting period, after which it
can be finalized — or cancelled at any point before that.

Each step is an intent builder: ``request``, ``cancel`` and ``finalize``
capture what to do, a route pin (``gasless()``, ``self_funded()``) narrows
how, and a terminal (``send()``, ``sign()``, ``verify()``) does it. Unpinned,
``send()`` prefers the gasless route and falls back to the user's own
transaction. Unlike a deposit, gasless withdrawal works for ETH too:
Core4Mica verifies the signature itself rather than leaning on what the
asset implements.

Finalization is the exception: it offers no ``sign()``, because it needs no
signature at all — the payout goes to the user and the amount was fixed when
they requested it. Port of ``sdk-rust/src/client/withdraw.rs``.
"""

from __future__ import annotations

from typing import Any, Dict, Optional, Union

from ..authorizations import (
    WithdrawalCancelAuthorization,
    WithdrawalRequestAuthorization,
)
from ..errors import AmountZeroError, InvalidParamsError, OutcomeUnknownError
from ..models import TxReceiptWaitOptions
from . import sig
from .ctx import ClientCtx
from .facilitator import (
    NAMES_THE_REQUEST,
    confirm_facilitator_echo,
    rejection_error,
    sponsorship_unavailable,
)
from .model import Asset, Route, WithdrawReceipt


class WithdrawClient:
    def __init__(self, ctx: ClientCtx) -> None:
        self._ctx = ctx

    def is_gasless_available(self) -> bool:
        return self._ctx.facilitator.is_configured()

    def request(self, asset: Union[Asset, str, None], amount: int) -> "RequestBuilder":
        """Starts a withdrawal request for *amount* of *asset*. The grace
        period runs from the request; finalize once it elapses."""
        if int(amount) <= 0:
            raise AmountZeroError("withdrawal amount must be positive")
        return RequestBuilder(self._ctx, Asset.coerce(asset), int(amount))

    def cancel(self, asset: Union[Asset, str, None]) -> "CancelBuilder":
        """Starts a cancellation of the pending withdrawal request for *asset*."""
        return CancelBuilder(self._ctx, Asset.coerce(asset))

    def finalize(self, asset: Union[Asset, str, None]) -> "FinalizeBuilder":
        """Starts a finalization of the elapsed withdrawal request for *asset*."""
        return FinalizeBuilder(self._ctx, Asset.coerce(asset))


# --- request -------------------------------------------------------------


class RequestBuilder:
    def __init__(self, ctx: ClientCtx, asset: Asset, amount: int) -> None:
        self._ctx = ctx
        self._asset = asset
        self._amount = amount

    def gasless(self) -> "GaslessRequest":
        """Pins the gasless route: the facilitator submits and pays, with no
        self-funded fallback."""
        return GaslessRequest(self._ctx, self._asset, self._amount)

    def self_funded(self) -> "SelfFundedRequest":
        """Pins the user's own transaction."""
        return SelfFundedRequest(self._ctx, self._asset, self._amount)

    async def send(
        self, wait_options: Optional[TxReceiptWaitOptions] = None
    ) -> WithdrawReceipt:
        """Requests the withdrawal, gaslessly where possible. A rejection that
        names the request itself is returned rather than retried, and so is an
        unknown outcome: the facilitator may already have submitted, and
        requesting again would overwrite it and restart the grace period."""
        if not self._ctx.facilitator.is_configured():
            return await self.self_funded().send(wait_options)
        try:
            return await self.gasless().send()
        except Exception as exc:
            if sponsorship_unavailable(exc, NAMES_THE_REQUEST):
                return await self.self_funded().send(wait_options)
            raise


class GaslessRequest:
    def __init__(self, ctx: ClientCtx, asset: Asset, amount: int) -> None:
        self._ctx = ctx
        self._asset = asset
        self._amount = amount

    async def sign(self) -> WithdrawalRequestAuthorization:
        """Signs the request without submitting it. Redeem by attaching it to
        a fresh builder:
        ``withdraw.request(asset, amount).gasless().authorization(auth).send()``."""
        return await sig.request_withdrawal_authorization(
            self._ctx, self._asset.address, self._amount
        )

    def authorization(
        self, authorization: WithdrawalRequestAuthorization
    ) -> "AuthorizedRequest":
        """Attaches a request authorization signed elsewhere."""
        return AuthorizedRequest(self._ctx, self._asset, self._amount, authorization)

    async def send(self) -> WithdrawReceipt:
        """Requests the withdrawal gaslessly. The user needs no native balance
        and makes no transaction."""
        authorization = await sig.request_withdrawal_authorization(
            self._ctx, self._asset.address, self._amount
        )
        return await _submit(
            self._ctx,
            {"action": "request", "authorization": authorization.to_payload()},
            authorization.user,
            authorization.asset,
        )


class AuthorizedRequest:
    def __init__(
        self,
        ctx: ClientCtx,
        asset: Asset,
        amount: int,
        auth: WithdrawalRequestAuthorization,
    ) -> None:
        self._ctx = ctx
        self._asset = asset
        self._amount = amount
        self._auth = auth

    def _checked(self) -> None:
        """The authorization names its asset and amount, so a builder that
        disagrees with it would submit terms the caller never stated."""
        if (
            self._auth.asset.lower() != self._asset.address.lower()
            or self._auth.amount != self._amount
        ):
            raise InvalidParamsError(
                f"authorization signs {self._auth.amount} of {self._auth.asset}, "
                f"but the builder asks {self._amount} of {self._asset.address}"
            )

    async def verify(self) -> None:
        """Preflight: runs every check a real submission would run, without
        spending anyone's gas."""
        self._checked()
        await _verify(
            self._ctx, {"action": "request", "authorization": self._auth.to_payload()}
        )

    async def send(self) -> WithdrawReceipt:
        """Requests the withdrawal with the attached authorization. The
        submitter needs no signer of their own."""
        self._checked()
        return await _submit(
            self._ctx,
            {"action": "request", "authorization": self._auth.to_payload()},
            self._auth.user,
            self._auth.asset,
        )


class SelfFundedRequest:
    def __init__(self, ctx: ClientCtx, asset: Asset, amount: int) -> None:
        self._ctx = ctx
        self._asset = asset
        self._amount = amount

    async def send(
        self, wait_options: Optional[TxReceiptWaitOptions] = None
    ) -> WithdrawReceipt:
        gateway = await self._ctx.gateway()
        receipt = await gateway.request_withdrawal(
            self._amount, self._asset.erc20_token, wait_options=wait_options
        )
        return _self_funded_receipt(self._ctx, receipt, self._asset)


# --- cancel --------------------------------------------------------------


class CancelBuilder:
    def __init__(self, ctx: ClientCtx, asset: Asset) -> None:
        self._ctx = ctx
        self._asset = asset

    def gasless(self) -> "GaslessCancel":
        return GaslessCancel(self._ctx, self._asset)

    def self_funded(self) -> "SelfFundedCancel":
        return SelfFundedCancel(self._ctx, self._asset)

    async def send(
        self, wait_options: Optional[TxReceiptWaitOptions] = None
    ) -> WithdrawReceipt:
        """Cancels the pending withdrawal request, gaslessly where possible."""
        if not self._ctx.facilitator.is_configured():
            return await self.self_funded().send(wait_options)
        try:
            return await self.gasless().send()
        except Exception as exc:
            if sponsorship_unavailable(exc, NAMES_THE_REQUEST):
                return await self.self_funded().send(wait_options)
            raise


class GaslessCancel:
    def __init__(self, ctx: ClientCtx, asset: Asset) -> None:
        self._ctx = ctx
        self._asset = asset

    async def sign(self) -> WithdrawalCancelAuthorization:
        """Signs the cancellation without submitting it."""
        return await sig.cancel_withdrawal_authorization(self._ctx, self._asset.address)

    def authorization(
        self, authorization: WithdrawalCancelAuthorization
    ) -> "AuthorizedCancel":
        """Attaches a cancellation authorization signed elsewhere."""
        return AuthorizedCancel(self._ctx, self._asset, authorization)

    async def send(self) -> WithdrawReceipt:
        """Cancels the pending withdrawal request gaslessly."""
        authorization = await sig.cancel_withdrawal_authorization(
            self._ctx, self._asset.address
        )
        return await _submit(
            self._ctx,
            {"action": "cancel", "authorization": authorization.to_payload()},
            authorization.user,
            authorization.asset,
        )


class AuthorizedCancel:
    def __init__(
        self, ctx: ClientCtx, asset: Asset, auth: WithdrawalCancelAuthorization
    ) -> None:
        self._ctx = ctx
        self._asset = asset
        self._auth = auth

    def _checked(self) -> None:
        if self._auth.asset.lower() != self._asset.address.lower():
            raise InvalidParamsError(
                f"authorization cancels for {self._auth.asset}, but the builder "
                f"asks {self._asset.address}"
            )

    async def verify(self) -> None:
        self._checked()
        await _verify(
            self._ctx, {"action": "cancel", "authorization": self._auth.to_payload()}
        )

    async def send(self) -> WithdrawReceipt:
        self._checked()
        return await _submit(
            self._ctx,
            {"action": "cancel", "authorization": self._auth.to_payload()},
            self._auth.user,
            self._auth.asset,
        )


class SelfFundedCancel:
    def __init__(self, ctx: ClientCtx, asset: Asset) -> None:
        self._ctx = ctx
        self._asset = asset

    async def send(
        self, wait_options: Optional[TxReceiptWaitOptions] = None
    ) -> WithdrawReceipt:
        gateway = await self._ctx.gateway()
        receipt = await gateway.cancel_withdrawal(
            self._asset.erc20_token, wait_options=wait_options
        )
        return _self_funded_receipt(self._ctx, receipt, self._asset)


# --- finalize ------------------------------------------------------------


class FinalizeBuilder:
    """No ``sign()`` on any route: ``finalizeWithdrawalFor`` is permissionless
    because it pays the user, so there is nothing to sign."""

    def __init__(self, ctx: ClientCtx, asset: Asset) -> None:
        self._ctx = ctx
        self._asset = asset

    def gasless(self) -> "GaslessFinalize":
        return GaslessFinalize(self._ctx, self._asset)

    def self_funded(self) -> "SelfFundedFinalize":
        return SelfFundedFinalize(self._ctx, self._asset)

    async def send(
        self, wait_options: Optional[TxReceiptWaitOptions] = None
    ) -> WithdrawReceipt:
        """Pays out the elapsed withdrawal request, gaslessly where possible."""
        if not self._ctx.facilitator.is_configured():
            return await self.self_funded().send(wait_options)
        try:
            return await self.gasless().send()
        except Exception as exc:
            if sponsorship_unavailable(exc, NAMES_THE_REQUEST):
                return await self.self_funded().send(wait_options)
            raise


class GaslessFinalize:
    def __init__(self, ctx: ClientCtx, asset: Asset) -> None:
        self._ctx = ctx
        self._asset = asset

    async def verify(self) -> None:
        """Preflight — worth more here than elsewhere, since finalization is
        the one step that can be refused purely by the clock."""
        await _verify(self._ctx, self._finalize_request())

    async def send(self) -> WithdrawReceipt:
        """Finalizes gaslessly. Takes no signature: ``finalizeWithdrawalFor``
        pays the user whoever submits it."""
        return await _submit(
            self._ctx,
            self._finalize_request(),
            self._ctx.signer_address,
            self._asset.address,
        )

    def _finalize_request(self) -> Dict[str, Any]:
        return {
            "action": "finalize",
            "user": self._ctx.signer_address,
            "asset": self._asset.address,
        }


class SelfFundedFinalize:
    def __init__(self, ctx: ClientCtx, asset: Asset) -> None:
        self._ctx = ctx
        self._asset = asset

    async def send(
        self, wait_options: Optional[TxReceiptWaitOptions] = None
    ) -> WithdrawReceipt:
        gateway = await self._ctx.gateway()
        receipt = await gateway.finalize_withdrawal(
            self._asset.erc20_token, wait_options=wait_options
        )
        return _self_funded_receipt(self._ctx, receipt, self._asset)


# --- shared --------------------------------------------------------------


def _self_funded_receipt(
    ctx: ClientCtx, receipt: Dict[str, Any], asset: Asset
) -> WithdrawReceipt:
    return WithdrawReceipt(
        tx_hash=receipt["transactionHash"],
        route=Route.SELF_FUNDED,
        account=ctx.signer_address,
        asset=asset.address,
        raw=receipt,
    )


async def _submit(
    ctx: ClientCtx, request: Dict[str, Any], user: str, asset: str
) -> WithdrawReceipt:
    response = await ctx.facilitator.post("withdraw", request)
    if not response.get("success"):
        raise rejection_error(response, response.get("error"))

    tx_hash = response.get("txHash")
    if not isinstance(tx_hash, str) or not tx_hash.startswith("0x"):
        raise OutcomeUnknownError("facilitator reported success without a txHash")

    return WithdrawReceipt(
        tx_hash=tx_hash,
        route=Route.GASLESS,
        account=confirm_facilitator_echo("user", response.get("user"), user),
        asset=confirm_facilitator_echo("asset", response.get("asset"), asset),
        network=response.get("network"),
        raw=response,
    )


async def _verify(ctx: ClientCtx, request: Dict[str, Any]) -> None:
    response = await ctx.facilitator.post("withdraw/verify", request)
    if response.get("isValid"):
        return
    raise rejection_error(response, response.get("invalidReason"))

File: packages/sdk/src/client/withdraw.ts
/**
 * Withdrawing collateral. A request opens a waiting period, after which it
 * can be finalized — or cancelled at any point before that.
 *
 * Each step is an intent builder: `request`, `cancel` and `finalize` capture
 * what to do, a route pin (`gasless()`, `selfFunded()`) narrows how, and a
 * terminal (`send()`, `sign()`, `verify()`) does it. Unpinned, `send()`
 * prefers the gasless route and falls back to the user's own transaction.
 * Unlike a deposit, gasless withdrawal works for ETH too: Core4Mica verifies
 * the signature itself rather than leaning on what the asset implements.
 *
 * Finalization is the exception: it offers no `sign()`, because it needs no
 * signature at all — the payout goes to the user and the amount was fixed
 * when they requested it. Port of `sdk-rust/src/client/withdraw.rs`.
 */

import type {
  WithdrawalCancelAuthorization,
  WithdrawalRequestAuthorization,
} from "@/authorizations";
import type { ClientCtx } from "@/client/ctx";
import {
  confirmFacilitatorEcho,
  NAMES_THE_REQUEST,
  rejectionError,
  sponsorshipUnavailable,
} from "@/client/facilitator";
import { Asset, Route, type WithdrawReceipt } from "@/client/model";
import * as sig from "@/client/sig";
import type { TxReceiptWaitOptions } from "@/contract";
import {
  AmountZeroError,
  InvalidParamsError,
  OutcomeUnknownError,
} from "@/errors";
import { parseU256 } from "@/utils";

export class WithdrawClient {
  constructor(private ctx: ClientCtx) {}

  isGaslessAvailable(): boolean {
    return this.ctx.facilitator.isConfigured();
  }

  /**
   * Start a withdrawal request for `amount` of `asset`. The grace period runs
   * from the request; finalize once it elapses.
   */
  request(
    asset: Asset | string | null | undefined,
    amount: number | bigint | string,
  ): RequestBuilder {
    const parsed = parseU256(amount);
    if (parsed <= 0n) {
      throw new AmountZeroError("withdrawal amount must be positive");
    }
    return new RequestBuilder(this.ctx, Asset.coerce(asset), parsed);
  }

  /** Start a cancellation of the pending withdrawal request for `asset`. */
  cancel(asset: Asset | string | null | undefined): CancelBuilder {
    return new CancelBuilder(this.ctx, Asset.coerce(asset));
  }

  /** Start a finalization of the elapsed withdrawal request for `asset`. */
  finalize(asset: Asset | string | null | undefined): FinalizeBuilder {
    return new FinalizeBuilder(this.ctx, Asset.coerce(asset));
  }
}

// --- request --------------------------------------------------------------

export class RequestBuilder {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
    private amount: bigint,
  ) {}

  /** Pin the gasless route: the facilitator submits and pays, with no self-funded fallback. */
  gasless(): GaslessRequest {
    return new GaslessRequest(this.ctx, this.asset, this.amount);
  }

  /** Pin the user's own transaction. */
  selfFunded(): SelfFundedRequest {
    return new SelfFundedRequest(this.ctx, this.asset, this.amount);
  }

  /**
   * Request the withdrawal, gaslessly where possible. A rejection that names
   * the request itself is returned rather than retried, and so is an unknown
   * outcome: the facilitator may already have submitted, and requesting again
   * would overwrite it and restart the grace period.
   */
  async send(waitOptions?: TxReceiptWaitOptions): Promise<WithdrawReceipt> {
    if (!this.ctx.facilitator.isConfigured()) {
      return this.selfFunded().send(waitOptions);
    }
    try {
      return await this.gasless().send();
    } catch (err) {
      if (sponsorshipUnavailable(err, NAMES_THE_REQUEST)) {
        return this.selfFunded().send(waitOptions);
      }
      throw err;
    }
  }
}

export class GaslessRequest {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
    private amount: bigint,
  ) {}

  /**
   * Sign the request without submitting it. Redeem by attaching it to a fresh
   * builder:
   * `withdraw.request(asset, amount).gasless().authorization(auth).send()`.
   */
  async sign(): Promise<WithdrawalRequestAuthorization> {
    return sig.requestWithdrawalAuthorization(
      this.ctx,
      this.asset.address,
      this.amount,
    );
  }

  /** Attach a request authorization signed elsewhere. */
  authorization(
    authorization: WithdrawalRequestAuthorization,
  ): AuthorizedRequest {
    return new AuthorizedRequest(
      this.ctx,
      this.asset,
      this.amount,
      authorization,
    );
  }

  /**
   * Request the withdrawal gaslessly. The user needs no native balance and
   * makes no transaction.
   */
  async send(): Promise<WithdrawReceipt> {
    const authorization = await sig.requestWithdrawalAuthorization(
      this.ctx,
      this.asset.address,
      this.amount,
    );
    return submit(
      this.ctx,
      { action: "request", authorization: authorization.toPayload() },
      authorization.user,
      authorization.asset,
    );
  }
}

export class AuthorizedRequest {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
    private amount: bigint,
    private auth: WithdrawalRequestAuthorization,
  ) {}

  /**
   * The authorization names its asset and amount, so a builder that disagrees
   * with it would submit terms the caller never stated.
   */
  private checked(): void {
    if (
      this.auth.asset.toLowerCase() !== this.asset.address.toLowerCase() ||
      this.auth.amount !== this.amount
    ) {
      throw new InvalidParamsError(
        `authorization signs ${this.auth.amount} of ${this.auth.asset}, ` +
          `but the builder asks ${this.amount} of ${this.asset.address}`,
      );
    }
  }

  /** Preflight: runs every check a real submission would run, without spending anyone's gas. */
  async verify(): Promise<void> {
    this.checked();
    await verifyRequest(this.ctx, {
      action: "request",
      authorization: this.auth.toPayload(),
    });
  }

  /**
   * Request the withdrawal with the attached authorization. The submitter
   * needs no signer of their own.
   */
  async send(): Promise<WithdrawReceipt> {
    this.checked();
    return submit(
      this.ctx,
      { action: "request", authorization: this.auth.toPayload() },
      this.auth.user,
      this.auth.asset,
    );
  }
}

export class SelfFundedRequest {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
    private amount: bigint,
  ) {}

  async send(waitOptions?: TxReceiptWaitOptions): Promise<WithdrawReceipt> {
    const gateway = await this.ctx.gateway();
    const receipt = await gateway.requestWithdrawal(
      this.amount,
      this.asset.erc20Token,
      waitOptions,
    );
    return selfFundedReceipt(this.ctx, receipt, this.asset);
  }
}

// --- cancel ---------------------------------------------------------------

export class CancelBuilder {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
  ) {}

  gasless(): GaslessCancel {
    return new GaslessCancel(this.ctx, this.asset);
  }

  selfFunded(): SelfFundedCancel {
    return new SelfFundedCancel(this.ctx, this.asset);
  }

  /** Cancel the pending withdrawal request, gaslessly where possible. */
  async send(waitOptions?: TxReceiptWaitOptions): Promise<WithdrawReceipt> {
    if (!this.ctx.facilitator.isConfigured()) {
      return this.selfFunded().send(waitOptions);
    }
    try {
      return await this.gasless().send();
    } catch (err) {
      if (sponsorshipUnavailable(err, NAMES_THE_REQUEST)) {
        return this.selfFunded().send(waitOptions);
      }
      throw err;
    }
  }
}

export class GaslessCancel {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
  ) {}

  /** Sign the cancellation without submitting it. */
  async sign(): Promise<WithdrawalCancelAuthorization> {
    return sig.cancelWithdrawalAuthorization(this.ctx, this.asset.address);
  }

  /** Attach a cancellation authorization signed elsewhere. */
  authorization(
    authorization: WithdrawalCancelAuthorization,
  ): AuthorizedCancel {
    return new AuthorizedCancel(this.ctx, this.asset, authorization);
  }

  /** Cancel the pending withdrawal request gaslessly. */
  async send(): Promise<WithdrawReceipt> {
    const authorization = await sig.cancelWithdrawalAuthorization(
      this.ctx,
      this.asset.address,
    );
    return submit(
      this.ctx,
      { action: "cancel", authorization: authorization.toPayload() },
      authorization.user,
      authorization.asset,
    );
  }
}

export class AuthorizedCancel {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
    private auth: WithdrawalCancelAuthorization,
  ) {}

  private checked(): void {
    if (this.auth.asset.toLowerCase() !== this.asset.address.toLowerCase()) {
      throw new InvalidParamsError(
        `authorization cancels for ${this.auth.asset}, but the builder ` +
          `asks ${this.asset.address}`,
      );
    }
  }

  async verify(): Promise<void> {
    this.checked();
    await verifyRequest(this.ctx, {
      action: "cancel",
      authorization: this.auth.toPayload(),
    });
  }

  async send(): Promise<WithdrawReceipt> {
    this.checked();
    return submit(
      this.ctx,
      { action: "cancel", authorization: this.auth.toPayload() },
      this.auth.user,
      this.auth.asset,
    );
  }
}

export class SelfFundedCancel {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
  ) {}

  async send(waitOptions?: TxReceiptWaitOptions): Promise<WithdrawReceipt> {
    const gateway = await this.ctx.gateway();
    const receipt = await gateway.cancelWithdrawal(
      this.asset.erc20Token,
      waitOptions,
    );
    return selfFundedReceipt(this.ctx, receipt, this.asset);
  }
}

// --- finalize -------------------------------------------------------------

/**
 * No `sign()` on any route: `finalizeWithdrawalFor` is permissionless because
 * it pays the user, so there is nothing to sign.
 */
export class FinalizeBuilder {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
  ) {}

  gasless(): GaslessFinalize {
    return new GaslessFinalize(this.ctx, this.asset);
  }

  selfFunded(): SelfFundedFinalize {
    return new SelfFundedFinalize(this.ctx, this.asset);
  }

  /** Pay out the elapsed withdrawal request, gaslessly where possible. */
  async send(waitOptions?: TxReceiptWaitOptions): Promise<WithdrawReceipt> {
    if (!this.ctx.facilitator.isConfigured()) {
      return this.selfFunded().send(waitOptions);
    }
    try {
      return await this.gasless().send();
    } catch (err) {
      if (sponsorshipUnavailable(err, NAMES_THE_REQUEST)) {
        return this.selfFunded().send(waitOptions);
      }
      throw err;
    }
  }
}

export class GaslessFinalize {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
  ) {}

  /**
   * Preflight — worth more here than elsewhere, since finalization is the one
   * step that can be refused purely by the clock.
   */
  async verify(): Promise<void> {
    await verifyRequest(this.ctx, this.finalizeRequest());
  }

  /**
   * Finalize gaslessly. Takes no signature: `finalizeWithdrawalFor` pays the
   * user whoever submits it.
   */
  async send(): Promise<WithdrawReceipt> {
    return submit(
      this.ctx,
      this.finalizeRequest(),
      this.ctx.signerAddress,
      this.asset.address,
    );
  }

  private finalizeRequest(): Record<string, unknown> {
    return {
      action: "finalize",
      user: this.ctx.signerAddress,
      asset: this.asset.address,
    };
  }
}

export class SelfFundedFinalize {
  constructor(
    private ctx: ClientCtx,
    private asset: Asset,
  ) {}

  async send(waitOptions?: TxReceiptWaitOptions): Promise<WithdrawReceipt> {
    const gateway = await this.ctx.gateway();
    const receipt = await gateway.finalizeWithdrawal(
      this.asset.erc20Token,
      waitOptions,
    );
    return selfFundedReceipt(this.ctx, receipt, this.asset);
  }
}

// --- shared ---------------------------------------------------------------

function selfFundedReceipt(
  ctx: ClientCtx,
  receipt: Extract<WithdrawReceipt["raw"], { transactionHash: unknown }>,
  asset: Asset,
): WithdrawReceipt {
  return {
    txHash: receipt.transactionHash,
    route: Route.SelfFunded,
    account: ctx.signerAddress,
    asset: asset.address,
    raw: receipt,
  };
}

async function submit(
  ctx: ClientCtx,
  request: Record<string, unknown>,
  user: string,
  asset: string,
): Promise<WithdrawReceipt> {
  const response = await ctx.facilitator.post("withdraw", request);
  if (!response.success) {
    throw rejectionError(response, response.error);
  }

  const txHash = response.txHash;
  if (typeof txHash !== "string" || !txHash.startsWith("0x")) {
    throw new OutcomeUnknownError(
      "facilitator reported success without a txHash",
    );
  }

  return {
    txHash,
    route: Route.Gasless,
    account: confirmFacilitatorEcho("user", response.user, user),
    asset: confirmFacilitatorEcho("asset", response.asset, asset),
    network:
      typeof response.network === "string" ? response.network : undefined,
    raw: response,
  };
}

async function verifyRequest(
  ctx: ClientCtx,
  request: Record<string, unknown>,
): Promise<void> {
  const response = await ctx.facilitator.post("withdraw/verify", request);
  if (response.isValid) {
    return;
  }
  throw rejectionError(response, response.invalidReason);
}
